### PR TITLE
wiseunpacker: init at 1.3.3

### PR DESCRIPTION
**Pull request by Gigahawk**
## Description of changes

A tool for extracting files out of Wiseinstaller files
https://github.com/mnadareski/WiseUnpacker/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix [hunk 1]
```diff
@@ -52,7 +52,26 @@ lib.makeOverridable(
...
+                  cd ..
+                fi
+              done
+            }
+
+            traverseRename
+          '';
           installPhase = ''
+            runHook preInstall
+
```
Comment (Gigahawk) [inline pkgs/build-support/dotnet/make-nuget-deps/default.nix:74]: Copied from https://github.com/NixOS/nixpkgs/blob/89f8fd07fb125f182ad5652eff777940e09aa5c2/pkgs/build-support/dotnet/fetchnuget/default.nix#L30-L46

Is there a reason there's two different fetchNuGet implementations that don't behave the same?

File: pkgs/tools/archivers/wiseunpacker/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{ fetchFromGitHub
+, buildDotnetModule
+, dotnetCorePackages
+, pkgs
+}:
+let
+  version = "1.3.3";
```
Comment (Gigahawk) [inline pkgs/tools/archivers/wiseunpacker/default.nix:7]: Latest release is 1.4.2, but seems to be broken. This release works fine for me.

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix (hunk 1 above)
Comment (corngood) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix:74]: The other one is a related to dotnet-packages, which is more or less obsolete.

Which package requires this?  I'd like to make sure we solve this once and for all based on how the filenames are escaped.

File: pkgs/tools/archivers/wiseunpacker/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  dotnetInstallFlags = [ "-f net8.0" ];
+
+  # Rename to something sensible
+  postFixup = ''
+    mv "$out/bin/Test" "$out/bin/WiseUnpacker"
+  '';
+
+  nugetDeps = ./deps.nix;
+
+  projectFile = "Test/Test.csproj";
```
Comment (corngood) [inline pkgs/tools/archivers/wiseunpacker/default.nix:32]: Should this be using the main project file, and specifying Test.csproj as `testProjectFile?

File: pkgs/tools/archivers/wiseunpacker/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    owner = "mnadareski";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-APbfo2D/p733AwNNByu5MvC9LA8WW4mAzq6t2w/YNrs=";
+  };
+
+  dotnet-sdk = dotnetCorePackages.sdk_8_0;
+  dotnet-runtime = dotnetCorePackages.runtime_8_0;
+
+  dotnetInstallFlags = [ "-f net8.0" ];
```
Comment (corngood) [inline pkgs/tools/archivers/wiseunpacker/default.nix:23]: If we did something like `dotnetFlags = "-p:TargetFrameworks=net8.0";`, would it cut down on the dependencies?

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix [hunk 5]
```diff
@@ -52,7 +52,26 @@ lib.makeOverridable(
             rm -rf .signature.p7s
           '';
 
+          preInstall = ''
```
Comment (corngood) [inline pkgs/build-support/dotnet/make-nuget-deps/default.nix:55]: This impacts a lot of packages, so it should be a separate commit.

**Review by corngood — CHANGES_REQUESTED** [review]

File: pkgs/tools/archivers/wiseunpacker/default.nix (hunk 4 above)
Comment (Gigahawk) [reply pkgs/tools/archivers/wiseunpacker/default.nix:23]: When I add this I get 
```
Executing dotnetConfigureHook
  Determining projects to restore...
/nix/store/1nmjsx4wkr7braiz5spq84b1ibvvz2jb-dotnet-sdk-8.0.303/sdk/8.0.303/NuGet.targets(169,5): error : Invalid restore input. Invalid target framework 'unsupported'. Input files: /build/source/Test/Test.csproj. [/build/source/Test/Test.csproj]
```

File: pkgs/tools/archivers/wiseunpacker/default.nix (hunk 3 above)
Comment (Gigahawk) [reply pkgs/tools/archivers/wiseunpacker/default.nix:32]: WiseUnpacker is technically a C# library, and comes with a test project that's just a CLI tool which exposes the basic functionality of extracting files out of a Wise packed installer. I don't think it would make sense to use the main project file since it's just a library with no executable.

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix (hunk 1 above)
Comment (Gigahawk) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix:74]: If I drop this change and rebuild I get:
```
CSC : error CS0006: Metadata file '/nix/store/c6728sp915jhl489f9k91jxhvcabh1sc-WiseUnpacker-1.3.3-nuget-deps/share/nuget/packages/system.valuetuple/4.5.0/ref/portable-net40+sl4+win8+wp8/System.ValueTuple.dll' could not be found [/build/source/WiseUnpacker/WiseUnpacker.csproj::TargetFramework=net40]
CSC : error CS0006: Metadata file '/nix/store/c6728sp915jhl489f9k91jxhvcabh1sc-WiseUnpacker-1.3.3-nuget-deps/share/nuget/packages/system.valuetuple/4.5.0/ref/portable-net40+sl4+win8+wp8/System.ValueTuple.dll' could not be found [/build/source/WiseUnpacker/WiseUnpacker.csproj::TargetFramework=net452]
```

When I look in the folder the right path shows up it's just URL encoded:
```
$ ls -l /nix/store/c6728sp915jhl489f9k91jxhvcabh1sc-WiseUnpacker-1.3.3-nuget-deps/share/nuget/packages/system.valuetuple/4.5.0/ref/
total 48
dr-xr-xr-x 2 root root 4096 Jan  1  1970 MonoAndroid10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 MonoTouch10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 net461
dr-xr-xr-x 2 root root 4096 Jan  1  1970 net47
dr-xr-xr-x 2 root root 4096 Jan  1  1970 netcoreapp2.0
dr-xr-xr-x 2 root root 4096 Jan  1  1970 netstandard2.0
dr-xr-xr-x 2 root root 4096 Jan  1  1970 portable-net40%2Bsl4%2Bwin8%2Bwp8
dr-xr-xr-x 2 root root 4096 Jan  1  1970 uap10.0.16299
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarinios10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarinmac20
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarintvos10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarinwatchos10
```

Comment (corngood) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix:74]: Could we:

- put this in `unpackPhase`
- give it a more specific name like `unescapeNupkgPaths`
- make the replacement logic more robust

I'm thinking something like:

`(e=${e/\\/\\\\} && echo -e ${e//%/\\x})`

It's only really old versions of the nuget client (before https://github.com/NuGet/NuGet.Client/pull/1855) that escaped the files like this, but the client still supports unescaping them, so we should too, when pulling binary packages.  We also unpack .nupkg files in other places, but they will all be packed with recent versions of the client.

Even with recent tools, filenames will be mangled if you intentionally put a file like `foo%2B` in a package.

File: pkgs/tools/archivers/wiseunpacker/default.nix (hunk 4 above)
Comment (corngood) [reply pkgs/tools/archivers/wiseunpacker/default.nix]: I just tried it, and I was able to fetch-deps without errors.  It failed during publish without `-f`, but I found that you can do: `-p:TargetFramework=net8.0` (singular), and it makes publish happy:

```
diff --git a/pkgs/by-name/wi/wiseunpacker/package.nix b/pkgs/by-name/wi/wiseunpacker/package.nix
index ef0584fe3463..f5d42fa68f9f 100644
--- a/pkgs/by-name/wi/wiseunpacker/package.nix
+++ b/pkgs/by-name/wi/wiseunpacker/package.nix
@@ -21,7 +21,7 @@ buildDotnetModule rec {
   dotnet-sdk = dotnetCorePackages.sdk_8_0;
   dotnet-runtime = dotnetCorePackages.runtime_8_0;
 
-  dotnetInstallFlags = [ "-f net8.0" ];
+  dotnetFlags = [ "-p:TargetFramework=net8.0" ];
 
   # Rename to something sensible
   postFixup = ''
```

This also removes almost all of the dependencies, including the one with the uri escaped paths.

**Review by corngood — CHANGES_REQUESTED** [review]

File: pkgs/tools/archivers/wiseunpacker/default.nix (hunk 4 above)
Comment (Gigahawk) [reply pkgs/tools/archivers/wiseunpacker/default.nix]: Ah ok sounds good, sorry I'm not that familiar with C#, but this change works on my end too.

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix (hunk 1 above)
Comment (Gigahawk) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix]: Sounds good, seems to work fine on my end, is there a convenient way to check if this breaks other packages?

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix [hunk 6]
```diff
@@ -53,6 +69,8 @@ lib.makeOverridable(
           '';
 
           installPhase = ''
+            runHook preInstall
```
Comment (corngood) [inline pkgs/build-support/dotnet/make-nuget-deps/default.nix:72]: This is a good idea, but we should really add `pre/postUnpack` and `postInstall` at the same time.  Can we put this in a separate PR for dotnet infra stuff?

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix [hunk 7]
```diff
@@ -41,6 +41,22 @@ lib.makeOverridable(
           unpackPhase = ''
             unzip -nq $src
             chmod -R +rw .
+
+            function unescapeNupkgPaths () {
```
Comment (corngood) [inline pkgs/build-support/dotnet/make-nuget-deps/default.nix:45]: Can we leave this for a separate PR?  It shouldn't be needed for `wiseunpacker` anymore, and it will require a lot of packages to be tested.

**Review by corngood — CHANGES_REQUESTED** [review]

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix (hunk 7 above)
Comment (Gigahawk) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix]: Sounds good, I'll move it to a new PR

**Review by corngood — APPROVED** [review]
Thanks!

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix (hunk 6 above)
Comment (Gigahawk) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix]: Moved to https://github.com/NixOS/nixpkgs/pull/337096

File: pkgs/build-support/dotnet/make-nuget-deps/default.nix (hunk 7 above)
Comment (Gigahawk) [reply pkgs/build-support/dotnet/make-nuget-deps/default.nix]: https://github.com/NixOS/nixpkgs/pull/337096

File: pkgs/tools/archivers/wiseunpacker/default.nix (hunk 2 above)
Comment (mnadareski) [reply pkgs/tools/archivers/wiseunpacker/default.nix]: Hi, original dev here. Nothing was ever reported to me about anything being broken. Please open an issue and describe what you're running into.

Comment (corngood) [reply pkgs/tools/archivers/wiseunpacker/default.nix]: I upgraded it locally to 1.4.2 and ran fetch-deps, and it builds okay.  There are no tests here though.  @Gigahawk do you remember what was wrong with it?

Comment (Gigahawk) [reply pkgs/tools/archivers/wiseunpacker/default.nix]: https://github.com/mnadareski/WiseUnpacker/issues/5
